File: repos/openrappter/python/openrappter/rappids/verify.py
"""Proof, not assertion.

Mirrors ``typescript/src/rappids/verify.ts``.

Every claim an organism makes about itself is re-derived here from bytes on
disk: the manifest hash over canonical JSON, the identity seed over the
traits, the identity motif over the RAPPID, each asset's exact length and
digest, and the body-frame chain. A creature that says it weighs three
megabytes and cannot produce them is not heavy, it is wrong -- and the
difference has to be visible in the report rather than smoothed over.

Nothing here is a heuristic and nothing is estimated. A check either passes
with the evidence that made it pass, or fails with what did not match.
"""

from __future__ import annotations

import os
import re
from pathlib import Path
from typing import List, Optional, Tuple

from .canonical import RAPP_EGG_DOMAIN, canonical_digest, rapp_hb, sha256_hex
from .identity import directory_hex, identity_drift, parse_rappid
from .store import (
    asset_bytes,
    asset_exists,
    body_frame_problems,
    read_rapp_object,
    resolve_within,
)
from .types import (
    AssetRecord,
    AssetVerification,
    BodyFrame,
    LoadedOrganism,
    QuantumRappidError,
    SonicProfile,
    VerificationCheck,
    VerificationReport,
)

_HEX64 = re.compile(r"^[0-9a-f]{64}$")


def _pass(name: str, detail: str) -> VerificationCheck:
    return VerificationCheck(name=name, status="pass", detail=detail)


def _fail(name: str, detail: str) -> VerificationCheck:
    return VerificationCheck(name=name, status="fail", detail=detail)


def _verify_asset(
    organism: LoadedOrganism, dimension: str, asset: AssetRecord
) -> AssetVerification:
    """One asset, weighed and hashed."""
    if not asset_exists(organism, dimension, asset.path):
        return AssetVerification(
            dimension=dimension,
            path=asset.path,
            status="missing",
            address_space=RAPP_EGG_DOMAIN,
            address_hash="",
            expected_bytes=asset.bytes,
            actual_bytes=None,
            expected_sha256=asset.sha256,
            actual_sha256=None,
            media_type=asset.media_type,
        )
    payload = asset_bytes(organism, dimension, asset.path)
    digest = sha256_hex(payload)
    if len(payload) != asset.bytes:
        status = "byte-mismatch"
    elif digest != asset.sha256:
        status = "hash-mismatch"
    else:
        status = "verified"
    return AssetVerification(
        dimension=dimension,
        path=asset.path,
        status=status,
        address_space=RAPP_EGG_DOMAIN,
        address_hash=rapp_hb(RAPP_EGG_DOMAIN, payload),
        expected_bytes=asset.bytes,
        actual_bytes=len(payload),
        expected_sha256=asset.sha256,
        actual_sha256=digest,
        media_type=asset.media_type,
    )


def _verify_frame_media(organism: LoadedOrganism) -> List[AssetVerification]:
    """Every media ref a frame carries, re-read from the local egg store.

    A frame names bytes by their RAPP/1 address. The address is only worth
    something if the bytes behind it are here and still hash to it, so each ref
    is resolved and re-hashed rather than trusted.
    """
    results: List[AssetVerification] = []
    for frame in organism.frames:
        media = frame.payload.get("media")
        if not isinstance(media, dict):
            # A frame whose media is not an object has already been named by
            # the chain check ("media is not an object"), and the report
            # carrying that failure is where it belongs. There are no refs here
            # to weigh, and inventing one would put a second, vaguer complaint
            # about the same frame in the asset list.
            continue
        for role, raw in media.items():
            dimension = str(frame.payload.get("dimension"))
            if not isinstance(raw, dict):
                results.append(
                    AssetVerification(
                        dimension=dimension,
                        path=role,
                        status="missing",
                        address_space=RAPP_EGG_DOMAIN,
                        address_hash="",
                        expected_bytes=0,
                        actual_bytes=None,
                        expected_sha256="",
                        actual_sha256=None,
                        media_type="application/octet-stream",
                    )
                )
                continue
            reference = raw.get("hash")
            address = reference if isinstance(reference, str) else ""
            expected_bytes = raw.get("bytes")
            expected_bytes = (
                expected_bytes
                if isinstance(expected_bytes, int) and not isinstance(expected_bytes, bool)
                else 0
            )
            media_type = raw.get("media_type")
            media_type = (
                media_type if isinstance(media_type, str) else "application/octet-stream"
            )
            payload = (
                read_rapp_object(organism, address) if _HEX64.fullmatch(address) else None
            )
            actual_hash = None if payload is None else rapp_hb(RAPP_EGG_DOMAIN, payload)
            if payload is None:
                status = "missing"
            elif len(payload) != expected_bytes:
                status = "byte-mismatch"
            elif raw.get("space") != RAPP_EGG_DOMAIN or actual_hash != address:
                status = "hash-mismatch"
            else:
                status = "verified"
            results.append(
                AssetVerification(
                    dimension=dimension,
                    path=role,
                    status=status,
                    address_space=RAPP_EGG_DOMAIN,
                    address_hash=address,
                    expected_bytes=expected_bytes,
                    actual_bytes=None if payload is None else len(payload),
                    expected_sha256=address,
                    actual_sha256=actual_hash,
                    media_type=media_type,
                )
            )
    return results


def _identity_checks(organism: LoadedOrganism) -> List[VerificationCheck]:
    checks: List[VerificationCheck] = []
    rappid = organism.document.rappid
    parts = parse_rappid(rappid)
    checks.append(_pass("identity.format", f"{rappid} parses as rappid:@owner/name:<64 hex>"))

    folder = os.path.basename(organism.directory.rstrip(os.sep))
    claimed = directory_hex(folder)
    if claimed is None:
        checks.append(
            _pass(
                "identity.habitat",
                f"{folder} is a named habitat directory, not an identity claim",
            )
        )
    elif claimed == parts.hex:
        checks.append(
            _pass("identity.habitat", "habitat directory matches the RAPPID tail-derived hex")
        )
    else:
        checks.append(
            _fail(
                "identity.habitat",
                f"habitat directory {folder} does not match the RAPPID hex {parts.hex}",
            )
        )

    claims: List[Tuple[str, Optional[str]]] = [("traits.json", organism.traits.rappid)]
    if organism.sonic is not None:
        claims.append(("sonic/sonic-profile.json", organism.sonic.rappid))
    for frame in organism.frames:
        claims.append((f"frames/{frame.seq:06d}.json", frame.stream_id))
    drift = identity_drift(rappid, claims)
    if drift:
        detail = "; ".join(f"{source} says {value}" for source, value in drift)
        checks.append(_fail("identity.single", f"identity drift: {detail}"))
    else:
        checks.append(
            _pass("identity.single", f"{len(claims) + 1} documents carry one identity")
        )

    parent = organism.document.parent_rappid
    checks.append(
        _pass("identity.lineage", "no parent pointer: this organism was minted, not born")
        if parent is None
        else _pass("identity.lineage", f"true offspring of {parent}")
    )
    return checks


def _manifest_check(sonic: SonicProfile) -> VerificationCheck:
    """Has the manifest been edited since it was written?

    A sidecar hashes the file bytes; an embedded ``manifest_sha256`` hashes the
    canonical JSON of every other key. Both are accepted, the sidecar first
    because it is the newer spelling. A profile carrying neither is not
    "probably fine": there is nothing to check it against, and a dimension
    whose integrity cannot be established must not read as verified.
    """
    if sonic.sidecar_sha256 is not None:
        if sonic.sidecar_sha256 == sonic.file_sha256:
            return _pass(
                "sonic.manifest",
                f"sonic-profile.sha256 matches the profile bytes ({sonic.file_sha256[:12]})",
            )
        return _fail(
            "sonic.manifest",
            f"sonic-profile.sha256 records {sonic.sidecar_sha256} but the profile hashes "
            f"to {sonic.file_sha256}",
        )
    if sonic.manifest_sha256 is not None:
        without_hash = {
            key: value for key, value in sonic.raw.items() if key != "manifest_sha256"
        }
        recomputed = canonical_digest(without_hash)
        if recomputed == sonic.manifest_sha256:
            return _pass(
                "sonic.manifest", f"manifest hash {recomputed[:12]} covers the profile"
            )
        return _fail(
            "sonic.manifest",
            f"manifest hash is {sonic.manifest_sha256} but the profile hashes to {recomputed}",
        )
    return _fail(
        "sonic.manifest",
        "the sonic profile records no manifest hash and has no sonic-profile.sha256 beside "
        "it, so nothing can establish that it has not been edited",
    )


def _midi_dna_check(sonic: SonicProfile) -> VerificationCheck:
    """The MIDI DNA is checked structurally, not against this runtime's generator.

    Whether a recorded motif is the one *this* implementation would derive is a
    question about a provider, and providers are replaceable by design -- the
    seam is the provider, not the RAPPID. Asserting our own melody over someone
    else's organism would turn a legitimate creature into a failing one. What
    must hold for any organism is that the motif is a well-formed 16-note
    ``NOTE(pitch, delta_onset, duration, velocity)`` sequence in MIDI range.
    """
    if len(sonic.prompt) != 16:
        return _fail(
            "sonic.midi-dna",
            f"the identity motif carries {len(sonic.prompt)} notes, expected 16",
        )
    bad = [
        note
        for note in sonic.prompt
        if not (0 <= note.pitch <= 127)
        or not (1 <= note.velocity <= 127)
        or note.delta_onset < 0
        or note.duration <= 0
    ]
    if bad:
        return _fail(
            "sonic.midi-dna", f"{len(bad)} of 16 identity-motif notes are out of MIDI range"
        )
    return _pass("sonic.midi-dna", "the 16-note identity motif is well formed and in MIDI range")


def _sonic_checks(organism: LoadedOrganism) -> List[VerificationCheck]:
    sonic = organism.sonic
    if sonic is None:
        return []
    checks: List[VerificationCheck] = [_manifest_check(sonic), _midi_dna_check(sonic)]

    trait_drift = [
        key for key, value in sonic.traits.items() if organism.traits.traits.get(key) != value
    ]
    checks.append(
        _pass("sonic.traits", "the sonic profile carries the same traits as traits.json")
        if not trait_drift
        else _fail(
            "sonic.traits",
            "traits disagree between traits.json and the sonic profile: "
            + ", ".join(sorted(trait_drift)),
        )
    )
    return checks


def _frame_checks(organism: LoadedOrganism) -> List[VerificationCheck]:
    if not organism.frames:
        return [
            _pass("frames.chain", "no body frames yet: a compact organism is not a broken one")
        ]
    problems: List[str] = []
    head: Optional[BodyFrame] = None
    for frame in organism.frames:
        for problem in body_frame_problems(frame, head, organism.document.rappid):
            problems.append(f"frame {frame.seq}: {problem}")
        head = frame
    if problems:
        return [_fail("frames.chain", "; ".join(problems))]
    return [
        _pass("frames.chain", f"{len(organism.frames)} append-only frames chain cleanly")
    ]


def _dimension_ref_checks(organism: LoadedOrganism) -> List[VerificationCheck]:
    """Dimension refs that name a file must name a file that is there.

    A ref without a separator is a cursor or an identifier (``"0002"``), not a
    path, and inventing a file for it would turn a healthy organism into a
    failing one.
    """
    missing: List[str] = []
    checked = 0
    for dimension in organism.document.dimensions:
        for key in sorted(dimension.refs):
            ref = dimension.refs[key]
            if "/" not in ref or "://" in ref:
                continue
            checked += 1
            try:
                target = resolve_within(organism.directory, ref)
            except QuantumRappidError as error:
                missing.append(f"{dimension.name}.{key} -> {error.message}")
                continue
            if not Path(target).exists():
                missing.append(f"{dimension.name}.{key} -> {ref}")
    if missing:
        return [
            _fail(
                "dimensions.refs",
                "dimension refs point at files that are not here: " + ", ".join(missing),
            )
        ]
    return [_pass("dimensions.refs", f"{checked} dimension refs resolve inside the organism")]


def verify_organism(organism: LoadedOrganism) -> VerificationReport:
    """Everything, checked.

    ``verified_addresses`` is deliberately a ``(dimension, sha256)`` set: it is
    what makes weight honest later, because the same bytes carried twice must
    not make an organism heavier.
    """
    checks: List[VerificationCheck] = [
        *_identity_checks(organism),
        *_sonic_checks(organism),
        *_frame_checks(organism),
        *_dimension_ref_checks(organism),
    ]

    assets: List[AssetVerification] = []
    if organism.sonic is not None:
        for asset in organism.sonic.assets:
            assets.append(_verify_asset(organism, organism.sonic.dimension, asset))
    assets.extend(_verify_frame_media(organism))

    broken = [asset for asset in assets if asset.status != "verified"]
    if broken:
        checks.append(
            _fail(
                "assets.content",
                "; ".join(f"{item.dimension}/{item.path} is {item.status}" for item in broken),
            )
        )
    else:
        checks.append(
            _pass("assets.content", f"{len(assets)} content addresses verified byte for byte")
        )

    verified_addresses = sorted(
        {
            f"{asset.address_space}:{asset.address_hash}"
            for asset in assets
            if asset.status == "verified"
        }
    )

    return VerificationReport(
        rappid=organism.document.rappid,
        ok=all(check.status == "pass" for check in checks),
        checks=checks,
        assets=assets,
        verified_addresses=verified_addresses,
    )


def is_verified(organism: LoadedOrganism) -> bool:
    """Sugar for the many call sites that only need the verdict."""
    return verify_organism(organism).ok


def assert_verified(organism: LoadedOrganism) -> VerificationReport:
    """Guard used before anything is appended to an organism."""
    report = verify_organism(organism)
    if not report.ok:
        failures = "; ".join(
            f"{check.name}: {check.detail}" for check in report.checks if check.status == "fail"
        )
        raise QuantumRappidError(
            "unverified", f"{organism.document.rappid} does not verify: {failures}"
        )
    return report

File: repos/openrappter/typescript/src/rappids/verify.ts
/**
 * Proof, not assertion.
 *
 * Every claim an organism makes about itself is re-derived here from bytes on
 * disk: the manifest hash over canonical JSON, the identity seed over the
 * traits, the identity motif over the RAPPID, each asset's exact length and
 * digest, and the body-frame chain. A creature that says it weighs three
 * megabytes and cannot produce them is not heavy, it is wrong — and the
 * difference has to be visible in the report rather than smoothed over.
 *
 * Nothing here is a heuristic and nothing is estimated. A check either passes
 * with the evidence that made it pass, or fails with what did not match.
 *
 * Mirrored by `python/openrappter/rappids/verify.py`.
 */

import {
  RAPP_EGG_DOMAIN,
  canonicalDigest,
  rappHb,
  sha256Hex,
} from './canonical.js';
import { directoryHex, identityDrift, parseRappid } from './identity.js';
import {
  assetBytes,
  assetExists,
  legacyBodyFrameProblems,
  readRappObject,
  resolveWithin,
} from './store.js';
import { QuantumRappidError } from './types.js';
import type {
  AssetRecord,
  AssetVerification,
  JsonObject,
  LoadedOrganism,
  VerificationCheck,
  VerificationReport,
} from './types.js';
import { basename, sep } from 'node:path';
import { existsSync } from 'node:fs';

function pass(name: string, detail: string): VerificationCheck {
  return { name, status: 'pass', detail };
}

function fail(name: string, detail: string): VerificationCheck {
  return { name, status: 'fail', detail };
}

/** One asset, weighed and hashed. */
function verifyAsset(
  organism: LoadedOrganism,
  dimension: string,
  asset: AssetRecord,
): AssetVerification {
  const base: Omit<AssetVerification, 'status' | 'actualBytes' | 'actualSha256'> = {
    dimension,
    path: asset.path,
    addressSpace: RAPP_EGG_DOMAIN,
    addressHash: '',
    expectedBytes: asset.bytes,
    expectedSha256: asset.sha256,
    mediaType: asset.mediaType,
  };
  if (!assetExists(organism, dimension, asset.path)) {
    return { ...base, status: 'missing', actualBytes: null, actualSha256: null };
  }
  const bytes = assetBytes(organism, dimension, asset.path);
  const digest = sha256Hex(bytes);
  base.addressHash = rappHb(RAPP_EGG_DOMAIN, bytes);
  if (bytes.length !== asset.bytes) {
    return { ...base, status: 'byte-mismatch', actualBytes: bytes.length, actualSha256: digest };
  }
  if (digest !== asset.sha256) {
    return { ...base, status: 'hash-mismatch', actualBytes: bytes.length, actualSha256: digest };
  }
  return { ...base, status: 'verified', actualBytes: bytes.length, actualSha256: digest };
}

function identityChecks(organism: LoadedOrganism): VerificationCheck[] {
  const checks: VerificationCheck[] = [];
  const rappid = organism.document.rappid;
  const parts = parseRappid(rappid);
  checks.push(pass('identity.format', `${rappid} parses as rappid:@owner/name:<64 hex>`));

  const folder = basename(organism.directory.replace(new RegExp(`${sep}+$`), ''));
  const claimed = directoryHex(folder);
  if (claimed === null) {
    checks.push(pass('identity.habitat', `${folder} is a named habitat directory, not an identity claim`));
  } else if (claimed === parts.hex) {
    checks.push(pass('identity.habitat', `habitat directory matches the RAPPID tail-derived hex`));
  } else {
    checks.push(
      fail('identity.habitat', `habitat directory ${folder} does not match the RAPPID hex ${parts.hex}`),
    );
  }

  const claims: Array<{ source: string; value: string | null }> = [
    { source: 'traits.json', value: organism.traits.rappid },
  ];
  if (organism.sonic !== null) claims.push({ source: 'sonic/sonic-profile.json', value: organism.sonic.rappid });
  for (const frame of organism.frames) {
    claims.push({
      source: `frames/${String(frame.seq).padStart(6, '0')}.json`,
      value: frame.stream_id,
    });
  }
  const drift = identityDrift(rappid, claims);
  checks.push(
    drift.length === 0
      ? pass('identity.single', `${claims.length + 1} documents carry one identity`)
      : fail(
          'identity.single',
          `identity drift: ${drift
            .map((entry) => `${entry.source} says ${String(entry.value)}`)
            .join('; ')}`,
        ),
  );

  const parent = organism.document.parentRappid;
  checks.push(
    parent === null
      ? pass('identity.lineage', 'no parent pointer: this organism was minted, not born')
      : pass('identity.lineage', `true offspring of ${parent}`),
  );
  return checks;
}

function sonicChecks(organism: LoadedOrganism): VerificationCheck[] {
  const sonic = organism.sonic;
  if (sonic === null) return [];
  const checks: VerificationCheck[] = [];

  checks.push(manifestCheck(sonic));
  checks.push(midiDnaCheck(sonic));

  const traitDrift = Object.keys(sonic.traits).filter(
    (key) => organism.traits.traits[key] !== sonic.traits[key],
  );
  checks.push(
    traitDrift.length === 0
      ? pass('sonic.traits', 'the sonic profile carries the same traits as traits.json')
      : fail('sonic.traits', `traits disagree between traits.json and the sonic profile: ${traitDrift.join(', ')}`),
  );
  return checks;
}

/**
 * Has the manifest been edited since it was written?
 *
 * A sidecar hashes the file bytes; an embedded `manifest_sha256` hashes the
 * canonical JSON of every other key. Both are accepted, the sidecar first
 * because it is the newer spelling. A profile carrying neither is not
 * "probably fine": there is nothing to check it against, and a dimension whose
 * integrity cannot be established must not read as verified.
 */
function manifestCheck(sonic: LoadedOrganism['sonic'] & object): VerificationCheck {
  if (sonic.sidecarSha256 !== null) {
    return sonic.sidecarSha256 === sonic.fileSha256
      ? pass('sonic.manifest', `sonic-profile.sha256 matches the profile bytes (${sonic.fileSha256.slice(0, 12)})`)
      : fail(
          'sonic.manifest',
          `sonic-profile.sha256 records ${sonic.sidecarSha256} but the profile hashes to ${sonic.fileSha256}`,
        );
  }
  if (sonic.manifestSha256 !== null) {
    const withoutHash: JsonObject = {};
    for (const key of Object.keys(sonic.raw)) {
      if (key !== 'manifest_sha256') withoutHash[key] = sonic.raw[key];
    }
    const recomputed = canonicalDigest(withoutHash);
    return recomputed === sonic.manifestSha256
      ? pass('sonic.manifest', `manifest hash ${recomputed.slice(0, 12)} covers the profile`)
      : fail(
          'sonic.manifest',
          `manifest hash is ${sonic.manifestSha256} but the profile hashes to ${recomputed}`,
        );
  }
  return fail(
    'sonic.manifest',
    'the sonic profile records no manifest hash and has no sonic-profile.sha256 beside it, '
      + 'so nothing can establish that it has not been edited',
  );
}

/**
 * The MIDI DNA is checked structurally, not against this runtime's generator.
 *
 * Whether a recorded motif is the one *this* implementation would derive is a
 * question about a provider, and providers are replaceable by design — the
 * seam is the provider, not the RAPPID. Asserting our own melody over someone
 * else's organism would turn a legitimate creature into a failing one. What
 * must hold for any organism is that the motif is a well-formed 16-note
 * `NOTE(pitch, delta_onset, duration, velocity)` sequence in MIDI range.
 */
function midiDnaCheck(sonic: LoadedOrganism['sonic'] & object): VerificationCheck {
  if (sonic.prompt.length !== 16) {
    return fail('sonic.midi-dna', `the identity motif carries ${sonic.prompt.length} notes, expected 16`);
  }
  const bad = sonic.prompt.filter(
    (note) =>
      note.pitch < 0
      || note.pitch > 127
      || note.velocity < 1
      || note.velocity > 127
      || note.deltaOnset < 0
      || note.duration <= 0,
  );
  return bad.length === 0
    ? pass('sonic.midi-dna', 'the 16-note identity motif is well formed and in MIDI range')
    : fail('sonic.midi-dna', `${bad.length} of 16 identity-motif notes are out of MIDI range`);
}

function frameChecks(organism: LoadedOrganism): VerificationCheck[] {
  if (organism.frames.length === 0) {
    return [pass(
      'frames.chain',
      'no legacy body.dimension frames yet: a compact organism is not broken',
    )];
  }
  const problems: string[] = [];
  let head: LoadedOrganism['frames'][number] | null = null;
  organism.frames.forEach((frame) => {
    for (const problem of legacyBodyFrameProblems(
      frame,
      head,
      organism.document.rappid,
    )) {
      problems.push(`frame ${frame.seq}: ${problem}`);
    }
    head = frame;
  });
  return [
    problems.length === 0
      ? pass(
        'frames.chain',
        `${organism.frames.length} legacy body.dimension frames pass integrity-only chaining; this is not current-kind conformance`,
      )
      : fail('frames.chain', problems.join('; ')),
  ];
}

function verifyFrameMedia(
  organism: LoadedOrganism,
): AssetVerification[] {
  const results: AssetVerification[] = [];
  for (const frame of organism.frames) {
    const media = frame.payload.media;
    for (const [role, raw] of Object.entries(media)) {
      if (raw === null || typeof raw !== 'object' || Array.isArray(raw)) {
        results.push({
          dimension: frame.payload.dimension,
          path: role,
          status: 'missing',
          addressSpace: RAPP_EGG_DOMAIN,
          addressHash: '',
          expectedBytes: 0,
          actualBytes: null,
          expectedSha256: '',
          actualSha256: null,
          mediaType: 'application/octet-stream',
        });
        continue;
      }
      const ref = raw as Record<string, unknown>;
      const hash = typeof ref.hash === 'string' ? ref.hash : '';
      const expectedBytes = typeof ref.bytes === 'number' ? ref.bytes : 0;
      const mediaType =
        typeof ref.media_type === 'string'
          ? ref.media_type
          : 'application/octet-stream';
      const bytes = /^[0-9a-f]{64}$/.test(hash)
        ? readRappObject(organism, hash)
        : null;
      const actualHash =
        bytes === null ? null : rappHb(RAPP_EGG_DOMAIN, bytes);
      let status: AssetVerification['status'] = 'verified';
      if (bytes === null) status = 'missing';
      else if (bytes.length !== expectedBytes) status = 'byte-mismatch';
      else if (
        ref.space !== RAPP_EGG_DOMAIN
        || actualHash !== hash
      ) {
        status = 'hash-mismatch';
      }
      results.push({
        dimension: frame.payload.dimension,
        path: role,
        status,
        addressSpace: RAPP_EGG_DOMAIN,
        addressHash: hash,
        expectedBytes,
        actualBytes: bytes?.length ?? null,
        expectedSha256: hash,
        actualSha256: actualHash,
        mediaType,
      });
    }
  }
  return results;
}

/**
 * Dimension refs that name a file must name a file that is there.
 *
 * A ref without a separator is a cursor or an identifier (`"0002"`), not a
 * path, and inventing a file for it would turn a healthy organism into a
 * failing one.
 */
function dimensionRefChecks(organism: LoadedOrganism): VerificationCheck[] {
  const missing: string[] = [];
  let checked = 0;
  for (const dimension of organism.document.dimensions) {
    for (const key of Object.keys(dimension.refs).sort()) {
      const ref = dimension.refs[key];
      if (!ref.includes('/') || ref.includes('://')) continue;
      checked += 1;
      let target: string;
      try {
        target = resolveWithin(organism.directory, ref);
      } catch (error) {
        missing.push(`${dimension.name}.${key} -> ${(error as QuantumRappidError).message}`);
        continue;
      }
      if (!existsSync(target)) missing.push(`${dimension.name}.${key} -> ${ref}`);
    }
  }
  return [
    missing.length === 0
      ? pass('dimensions.refs', `${checked} dimension refs resolve inside the organism`)
      : fail('dimensions.refs', `dimension refs point at files that are not here: ${missing.join(', ')}`),
  ];
}

/**
 * Everything, checked.
 *
 * `verifiedAddresses` is deliberately a `(dimension, sha256)` set: it is what
 * makes weight honest later, because the same bytes carried twice must not
 * make an organism heavier.
 */
export function verifyOrganism(organism: LoadedOrganism): VerificationReport {
  const checks: VerificationCheck[] = [
    ...identityChecks(organism),
    ...sonicChecks(organism),
    ...frameChecks(organism),
    ...dimensionRefChecks(organism),
  ];

  const assets: AssetVerification[] = [];
  if (organism.sonic !== null) {
    for (const asset of organism.sonic.assets) assets.push(verifyAsset(organism, organism.sonic.dimension, asset));
  }
  assets.push(...verifyFrameMedia(organism));

  const broken = assets.filter((asset) => asset.status !== 'verified');
  checks.push(
    broken.length === 0
      ? pass('assets.content', `${assets.length} content addresses verified byte for byte`)
      : fail(
          'assets.content',
          broken.map((asset) => `${asset.dimension}/${asset.path} is ${asset.status}`).join('; '),
        ),
  );

  const verifiedAddresses = [
    ...new Set(
      assets
        .filter((asset) => asset.status === 'verified')
        .map((asset) => `${asset.addressSpace}:${asset.addressHash}`),
    ),
  ].sort();

  return {
    rappid: organism.document.rappid,
    ok: checks.every((check) => check.status === 'pass'),
    checks,
    assets,
    verifiedAddresses,
  };
}

/** Sugar for the many call sites that only need the verdict. */
export function isVerified(organism: LoadedOrganism): boolean {
  return verifyOrganism(organism).ok;
}

/** Guard used before anything is appended to an organism. */
export function assertVerified(organism: LoadedOrganism): VerificationReport {
  const report = verifyOrganism(organism);
  if (!report.ok) {
    const failures = report.checks.filter((check) => check.status === 'fail');
    throw new QuantumRappidError(
      'unverified',
      `${organism.document.rappid} does not verify: ${failures.map((check) => `${check.name}: ${check.detail}`).join('; ')}`,
    );
  }
  return report;
}
